### DISCUSSION TITLE
When are we releasing v21?

### DISCUSSION THREAD

**Discussion by fballiano** [OpenMage/magento-lts#3654] [Releases]
The [list of changes is v21](https://github.com/openmage/magento-lts/#between-openmage-20x-and-21x-unreleased-available-on-branch-next) is:
- PHP 8.1 as minimum required version
- Removed scriptaculous/dragdrop.js (#3215)
- RWD theme: updated jQuery to 3.7.0 (#3204)
- Unified CSRF configuration (#3147) and added form key validation to Contacts form (#3146)
- Removed double span element from HTML buttons (#3123)
- Removed all deprecated Mysql4_ classes (#2730). If there are any old modules/extensions in your installation that use such classes, you must run shell/rename-mysql4-class-to-resource.php in the command line in order to convert them. Backup all files before running the script
- Removed "admin routing compatibility mode" (#1551)

PHP8.1 could be tricky for somebody but PHP 7.4 is EOL since a while and [PHP 8.0 is reaching EOL in a few days](https://www.php.net/supported-versions.php) so...

Yes, they are breaking changes and some of them will require work on the store's end but at the end of the day... either we release them or we can't leave them there forever if we don't want to release them ever.

I think that we have to address this issue and decide if and when we will release v21.

In https://github.com/OpenMage/rfcs/blob/main/accepted/0002-release-schedule.md it was approved to release a max of 2 major releases per year, but we're not releasing any since a few years.

What are your ideas for v21?

**Comment by Flyingmana** [comment]
probably depends a bit how long we offer support for the previous one.
end of the Year is for many the most busy time, where you dont want to move to many additional parts. Then Module Vendors might need some time to align for the new major version, where the interfaces for modules changed.

Generally I think releasing a new Version in the middle or end of Q1 is a good time.

**Comment by empiricompany** [comment]
I am very concerned about the minimum PHP 8 version. I have only updated a few shops, and it can be quite troublesome if there are many poorly written extensions (would it be helpful to have a guide for upgrading to PHP 8?). However, the world must move forward, and it certainly is an added value.

**Reply by kiatng** [reply]
In the process of upgrading a couple of sites to PHP8.1. It's a pain, high uncertainty if it'll break production. For a site without dedicated developers, I would say it's near impossible to upgrade with certainty. I think majority will stick with PHP7 for very long time.

**Reply by fballiano** [reply]
I didn't have nearly as many problems but it's for sure that it needs to be checked.

People stayed for many years with php5.6 but I wouldn't suggest it anyway and, as a platform, I think we should push for the safe upgrades, stores will have to pay for some developers but that's how it works, not everything can always be painless and free.

**Reply by alexh-swdev** [reply]
PHP7 is so long out of support.... I really don't think it makes any sense to A) support it any longer and B) use any software only supporting PHP7. I don't want be responsibly for security-related issues like break-ins jsut because of that.

Yes, the switch from 7 to 8 here producted some headaches beforehand. Not necessarily because of OM, but more because of unsopported modules. We even through away some closed-source ioncubed modules. But with the help of some scripts found here and there in the web and OM- and Server-logfiles, it was doable and not so hard as expected.

**Reply by addison74** [reply]
PHP7.x has been left unsupported for some time, but there are installations that still use it successfully in online stores with many products. Since I've been working with PHP (2000), I've never come across the notion of a "security breach", and fortunately the vulnerabilities have not been exploited. This is also because a server is not only in PHP. This does not mean that I do not want to use the latest version, but there is no rush. If I were to start a brand new store, I would choose everything that is new at this moment (Debian 12, PHP 8.3, ....) to stay quiet for a while, but I know that this is not possible. I have installations where team work is requested with many weeks to upgrade and the customers have started to be a bit tight on their wallets.

At this moment we'd better try to get rid of Prototype in the next year, which is a dangerous relic from the past, with the risk of staying for a while with PHP 7.4. I have been using Debian since it was released on all servers and I have rarely installed other operating systems. After seeing what happened recently with CentOS I realized that I made a good choice in the past. Debian 11 comes with an LTS support until 2026 and it was shipped with PHP 7.4.

**Comment by addison74** [comment]
Even if there is an RFC that sets rules when new versions can be released, we should discuss some important aspects, from the users' perspective.

Those who are still on OpenMage 19 for good reasons, either choose another platform, or evaluate the transition to version 20. As you all know, an upgrade comes with problems, smaller or bigger.

If we are moving to version 21, then those who are in version 20 will face issues, as it is known that there are a lot of BCs announced publicly and others undiscovered. These users are loyal to the platform, they made efforts to migrate.

I wonder why we don't have only 2 versions, one stable main and one test, in which we add various features that, if they prove stable and useful, should be ported into the main branch? There is a risk that we will stay less and less in the boat with the passage of time.

**Reply by Flyingmana** [reply]
its in part a different Topic...
But having only test and main still forces people to move forward in the same way as now.
The original idea of 19 was to give people time before they need to migrate to the followup version, and this time is soon going to be up, regardless of if we release 21. And that people on version 20 have to expect BC Breaks more early was already known with the initial announced lifetime of 1,5 years.

The longer we wait with a 21 Release, the more BC Breaks will accumulate, no?

**Reply by sreichel** [reply]
> I wonder why we don't have only 2 versions, one stable main and one test

We already had it with v1...

Do we need `next` branch at all? Think there are already some breaking changes in `v20`.

**Reply by addison74** [reply]
If I only use OpenMage, there is no problem to migrate from 19 to 20 to 21 and so on. Extensions/modules and custom code can create problems and the big dilemma is that we don't have feedback from anyone who will be affected by these changes. This year I brought two stores to v20, but it was with effort because of some extensions and changes that I didn't know about. Honestly speaking, I would not repeat the experience too soon. There have been several reports over the last few years, but not enough to evaluate all the BCs we create.

**Reply by fballiano** [reply]
in v21 there are only an extremely limited number of PRs, all documented and linked in the README, so there's no real surprise there. In v20 instead there have been many changes that were thought not to be breaking, but actually they were.

**Comment by fballiano** [comment]
February 2024 came, what are we doing with the next branch? I would like to release it as soon as possible and start moving on.
